File: 6/solution.ts
// Path: 6/solution.ts

export function part1(input: string): number {
    let index = 0;

    for (let i = 0; i < input.length; i++) {
        if (
            input[i] === input[i + 1] ||
            input[i] === input[i + 2] ||
            input[i] === input[i + 3] ||
            input[i + 1] === input[i + 2] ||
            input[i + 1] === input[i + 3] ||
            input[i + 2] === input[i + 3]
        ) {
            continue;
        } else {
            index = i + 4;
            break;
        }
    }

    return index;
}

export function part2(input: string): number {
    let index = 0;

    for (let i = 13; i < input.length; i++) {
        if (new Set(input.slice(i - 14, i)).size === 14) {
            index = i;
            break;
        }
    }

    return index;
}

const input = Deno.readTextFileSync("./6/input.txt");

console.log("Part 1:", part1(input));
console.log("Part 2:", part2(input));

File: 6/solution.rs
// Path: 6/solution.rs

use std::{fs::File, io::Read};

fn part1(input: &str) -> usize {
    input
        .as_bytes()
        .windows(4)
        .position(|w| {
            w.iter()
                .all(|&x| w.iter().filter(|&y| x == *y).count() == 1)
        })
        .unwrap()
        + 4
}

fn part2(input: &str) -> usize {
    input
        .as_bytes()
        .windows(14)
        .position(|w| {
            // all unique
            w.iter()
                .all(|&x| w.iter().filter(|&y| x == *y).count() == 1)
        })
        .unwrap()
        + 14
}

fn read_input() -> String {
    let mut input = String::new();
    let mut file = File::open("./6/input.txt").unwrap();
    file.read_to_string(&mut input).unwrap();
    input
}

fn main() {
    let input = read_input();
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
